Immutable block manager

braingram opened the pull request:
Not sure what happened to #1536 but it stopped accepting new commits. Reopening this test PR.

braingram commented:
Roman regression tests showing no new errors (19 are failing due to unrelated roman issues)
https://plwishmaster.stsci.edu:8081/blue/organizations/jenkins/RT%2FRoman-devdeps/detail/Roman-devdeps/286/pipeline/185

braingram commented:
Running the benchmarks on this PR vs main I see only one major difference. For 'large' files (these have 26 x 26 = 676 tree nodes) both `custom_tree_to_tagged_tree` and `tagged_tree_to_custom_tree` are roughly 2x slower with this PR. My initial suspicion is that this has to do with the addition of the `Serialization` and `Deserialization` operations that get created, used and discarded for each object. It probably makes more sense to have these created once for each tree operation and re-used/reset for each object.

braingram commented:
> Running the benchmarks on this PR vs main I see only one major difference. For 'large' files (these have 26 x 26 = 676 tree nodes) both `custom_tree_to_tagged_tree` and `tagged_tree_to_custom_tree` are roughly 2x slower with this PR. My initial suspicion is that this has to do with the addition of the `Serialization` and `Deserialization` operations that get created, used and discarded for each object. It probably makes more sense to have these created once for each tree operation and re-used/reset for each object.

The changes in [52f7bf8](https://github.com/asdf-format/asdf/pull/1537/commits/52f7bf8b16f04c44d784f5768d2c48f804365f25) and [77fb0e6](https://github.com/asdf-format/asdf/pull/1537/commits/77fb0e6655a65fd4f8d53ce230902148c20ca517) improve the performance of these operations by:
- indexing blocks to be written by their data for fast finding of shared blocks (40% improvement)
- create one `Serialization`/`Deserialization` per call to `custom_tree_to_tagged_tree`/`tagged_tree_to_custom_tree` and do not use it as a context manager (14% improvement)
- pre-calculate the context version string (to avoid a call to the `ctx.version_string` property for every node) (4% improvement)
- cache converter and AsdfType lookups (3% improvement)
In total this makes `custom_tree_to_tagged_tree` about 20% faster compared to main and `tagged_tree_to_custom_tree` is a few percent faster.

The JWST regression tests finished with 74 fewer errors than the previous run which suggests other changes are confounding the result:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-devdeps/662/
I ran a follow up using asdf main:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-devdeps/663/
that had 2 fewer errors:
- test_nirspec_missing_msa_nofail
- test_nirspec_missing_msa_fail
These will require some investigation. It appears the jwst downstream tests are also now failing although I don't immediately see how the failures are related to the recent changes.

braingram commented:
Comparing the jwst CI tests to a run on a PR that includes only documentation changes shows the same failures as the run here:
https://github.com/spacetelescope/jwst/actions/runs/4996866055?pr=7589
I think it's safe to say the 8 failing tests in jwst are not related to the changes in this asdf PR.

braingram commented:
I've been so far unable to replicate the 2 failed jwst regression tests.
I am rerunning the regression tests to see if they repeat. They only occurred in the mac-os run which appears stuck for the rerun I started earlier today:
https://plwishmaster.stsci.edu:8081/blue/organizations/jenkins/RT%2FJWST-devdeps/detail/JWST-devdeps/664/pipeline/192

braingram commented:
@eslavich while doing some testing with reading and writing to (a fake) s3. I encountered 2 new issues that are related to this PR.

All of the tests so far have been with a fake s3 so I need to verify that this is the same for real s3.

The first issue is that reading from s3 (using s3fs) produces a file-like object that has no valid fileno. This means that `np.fromfile` used in `GenericFile.read_into_array` will fail. This issue also impacts the code in main so I opened a separate issue: https://github.com/asdf-format/asdf/issues/1552

The second issue is that writing to s3 (using s3fs) produces a file-like object that is not seekable but in this case returns valid results for 'tell'. This differs from https://github.com/asdf-format/asdf/issues/1542 as it appears that os.pipe produces a file that is not seekable and calling `tell` raises an exception. The changes in this PR skipped attempting to write a block index to a non-seekable file but with this new case (for the file-like object produced by s3fs) we can and likely should write a block index. I'm inclined to fix this issue in this PR since it does not impact main. However I don't want this change to negatively impact any in-progress review of this PR.

Would you like me to open a separate issue (and follow-up PR) for the second issue above or update this PR to fix the issue?

braingram commented on asdf/_block/external.py:
```diff
@@ -0,0 +1,42 @@
+from asdf import generic_io, util
+
+
+class UseInternal:
```
Thanks! I updated the usage here:
228e02242289bdef100070acc27c93e18916d552

braingram commented on asdf/_block/options.py:
```diff
@@ -0,0 +1,69 @@
+
+    def __init__(self, storage_type=None, compression_type=None, compression_kwargs=None):
+        if storage_type is None:
+            with config_context() as cfg:
```
Good point. I switched it to `get_config` here: d7202ebbf5019a6fd4f8040fd73f1ff6e31011a9

braingram commented on asdf/_block/options.py:
```diff
@@ -0,0 +1,69 @@
+        self._compression_kwargs = kwargs
+
+    def __copy__(self):
+        return type(self)(self._storage_type, self._compression, self._compression_kwargs)
```
The two are equivalent here, I switched `Key` to use `type(self)` so they match: 1e49bd1b6c5649fd9bb9217ac58a87663fae0027

braingram commented on asdf/generic_io.py:
```diff
@@ -805,8 +780,10 @@ def memmap_array(self, offset, size):
             nbytes = self._fd.tell()
-            self._fd.seek(loc, 0)
             self._mmap = mmap.mmap(self._fd.fileno(), nbytes, access=acc)
+            # on windows mmap seeks to the start of the file so return the file
```
Agreed :)

braingram commented on asdf/_block/store.py:
```diff
@@ -0,0 +1,128 @@
+
+    def assign_object(self, obj, value):
+        if isinstance(obj, Key):
+            obj_id = id(obj._ref())
```
Good catch. I updated the code to use `Key._is_valid` (which checks for `_ref is None` and `_ref() is None`) before attempting to resolve the weak reference:
eb679e15a8c2b53130e5bea333b7e4e11cbbbf80

braingram commented on asdf/_block/store.py:
```diff
@@ -0,0 +1,128 @@
+            # we already have a key, check if it's already in the store
+            if obj_key in by_key:
+                by_key[obj_key] = value
+                return
```
Good eye :) I removed it: eb679e15a8c2b53130e5bea333b7e4e11cbbbf80

braingram commented on asdf/_block/writer.py:
```diff
@@ -0,0 +1,105 @@
+
+class WriteBlock:
+    """
+    Data and compression options needed to write and ASDF block.
```
Fixed in: 07c49ac940d7addc688046af92926ec2140f519f

braingram commented on asdf/_block/io.py:
```diff
@@ -0,0 +1,521 @@
+            return data
+
+        data = callback
+        fd.fast_forward(header["allocated_size"])
```
I modified this to `fast_forward` to the end of the file for a streamed block.

braingram commented on asdf/_block/external.py:
```diff
@@ -0,0 +1,42 @@
+
+class ExternalBlockCache:
+    def __init__(self):
+        self._cache = {}
```
I'm not quite sure about this one.

The old code called `AsdfFile.open_external` which caches every opened AsdfFile (in a dictionary). This means that a repeat call to `open_external` (and then accessing the first block) always returns the same array as `Block` caches the data). This behavior is relied upon for matching arrays that share an external block as tested in `test_explode_then_implode`:
https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/_tests/commands/tests/test_exploded.py#L11-L51
Changing `ExternalBlockCache` to LRU would mean that when the cache overflows, a call to `load` would return an array with a different id which could interfere with settings assignment and block sharing.

I'm inclined to view the memory mapping of external blocks as a bug (and fix it in this PR by not memory mapping) but that does create an issue for large exploded files as you pointed out. As `ExternalBlockCache` is used by the `Manger` that is aware of if the blocks should be lazy loaded and/or memory mapped another option would be to see how complicated it would be to pass these options to `ExternalBlockCache.load` and allow the user some control over how the external blocks are loaded. What do you think about this as an option for addressing the concern about large exploded files?

braingram replied:
I spent a bit of time looking at this today and I think there may be a few issues with passing along open options. 

`lazy_load` is not an issue (and is [currently supported in this PR](https://github.com/braingram/asdf/blob/642fd7f1c244d9d7662a5e5563e2a98f6f30e358/asdf/core/_converters/ndarray.py#L162-L181) as the ndarray converter sets up for lazy loading the external block then forces it's loading if `lazy_load` was disabled).

`validate_checksums` also should not be an issue but is not currently being passed on for this PR.

`copy_arrays` is where things get complicated. The main complication is when the external file is closed. For a non-memmapped external block, the external file can be opened, the block data read and the file closed before the block data is returned. This is not true for memmapped blocks where the external file will need to be held open until the `AsdfFile` instance is closed. This was one of the reasons why the previous `BlockManager` held a reference to the `AsdfFile` (so that it could used `AsdfFile.open_external` which can keep the external files open and close them when the `AsdfFile` is closed).

I could see a few options to make memory mapping external blocks possible but I'm not sure if the complexity outweighs the benefits.

1. Duplicate the `AsdfFile.open_external` functionality in `_block.Manager`. asdf prior to this PR called `close` on the block manager and we will need to re-add that call to allow the block manager to close the external files.
2. Allow for generating pure `numpy.memmap` objects (which don't guarantee file closing) for external blocks.

I'm going to think about this one some more but any suggestions on how to handle this are appreciated.

braingram commented on asdf/_block/key.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class Key:
+    _next = 0
+
+    @classmethod
+    def _next_key(cls):
+        key = cls._next
+        cls._next += 1
+        return key
```
I originally considered a uuid (where I learned they're not [guaranteed to be unique](https://en.wikipedia.org/w/index.php?title=Universally_unique_identifier&oldid=755882275#Random_UUID_probability_of_duplicates)?).

One consideration is performance. On my machine, running the following:
```python
import time
import uuid


class UUIDKey:
    def __init__(self):
        self._key = uuid.uuid4()


class NumKey:
    _next = 0
    def __init__(self):
        self._key = NumKey._next
        NumKey._next += 1

if __name__ == '__main__':
    n = 10000
    print(f"N={n}")
    for Key in (UUIDKey, NumKey):
        t0 = time.perf_counter()
        keys = [Key() for _ in range(n)]
        t1 = time.perf_counter()
        print(f"{Key.__class__.__name__}: {t1-t0}")
```

Outputs:
```
(asdf_3p0) aerum2022:tests bgraham$ python time_key.py
N=10000
type: 0.016155499999513268
type: 0.00325820800026122
```

The uuid takes ~5x longer.

As the `_key` usage is entirely internal it shouldn't be an issue to swap this out with a uuid if we can think of a situation where incrementing the number is an issue.

braingram commented on asdf/tags/core/stream.py:
```diff
@@ -0,0 +1,40 @@
+from .ndarray import NDArrayType, numpy_dtype_to_asdf_datatype
+
+
+class Stream(NDArrayType):
```
I don't think so (and removing the `NDArrayType` inheritance does not result in any test failures).

I'm not aware of any usage of `Stream` that relies upon this inheritance and some of the inherited methods do not work:
```python
s = asdf.Stream([1024], dtype='f8')
s.dtype   # raises AttributeError: 'NoneType' object has no attribute 'dtype'
```
I removed this in: 78cd8185677aa1da795de3393d78b1645500e924

braingram commented on asdf/_block/io.py:
```diff
@@ -0,0 +1,521 @@
+
+    Raises
+    ------
+    OSError
```
I agree. I picked this mainly due to existing usage in block io and to try and limit the number of changes (more on that below).

The previous code did not raise any error when attempts to read a block index failed: https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/block.py#L409

I think raising an error makes the most sense. The only usage of `read_block_index` is in `asdf._block.reader.read_blocks`: https://github.com/braingram/asdf/blob/78cd8185677aa1da795de3393d78b1645500e924/asdf/_block/reader.py#L222 where a failed attempt to read the block index will result in falling back to reading the blocks serially.

Using a custom exception seems like an improvement but the use of custom exceptions was not common in the old block code. I did not attempt to update this (as part of this PR) to use custom exceptions and instead tried to keep the errors consistent (both in type and message). In this situation (where there was no previous error) I opted to stick with the commonly used error type (which was originally IOError which python replaced with OSError) instead of implementing a custom exception.

Do you think a custom exception is called for here? If so, should it be public (perhaps simply adding it to `asdf.exceptions`)?

braingram replied:
I converted this to a custom private exception `BlockIndexError` (which does not inherit from `OSError`).
642fd7f1c244d9d7662a5e5563e2a98f6f30e358
Looking through the backlog of issues it looks like we already have one for applying this throughout asdf:
https://github.com/asdf-format/asdf/issues/1222

braingram commented on asdf/_block/io.py:
```diff
@@ -0,0 +1,521 @@
+        raise RuntimeError(msg)
+    header = BLOCK_HEADER.pack(**header_kwargs)
+    padding_bytes = header_kwargs["allocated_size"] - header_kwargs["used_size"]
+    return header, buff, padding_bytes
```
Thanks! I updated `generate_write_header` to instead return a dict: b6f53d6f50c7eb40d07ca2566d3663a89d668838

braingram commented on asdf/_block/external.py:
```diff
@@ -0,0 +1,42 @@
+def uri_for_index(uri, index):
+    parts = list(util.patched_urllib_parse.urlparse(uri))
+    path = parts[2]
+    dirname, filename = os.path.split(path)
```
`urlparse` returns a named tuple that contains a `path` component that is separated with os-native characters. I made a few changes to hopefully clarify this and renamed the function to `relative_uri_for_index`: 2ab79e81f28d06fab07dc1ddf4eeaa1ae4f693b1

Does this help to clarify the usage in the function?

braingram commented on asdf/_block/writer.py:
```diff
@@ -0,0 +1,105 @@
+        offsets.append(offset)
+        fd.write(constants.BLOCK_MAGIC)
+        headers.append(bio.write_block(fd, streamed_block.data_bytes, stream=True))
+    elif len(offsets) and write_index and fd.seekable():
```
Exactly!

While attempting to write a non-seekable file test I ran into this:
https://github.com/asdf-format/asdf/issues/1542
and:
https://github.com/asdf-format/asdf/issues/1545

I thought this could be mostly handled by not calling tell on a non-seekable file (which meant that no block index could be generated for a non-seekable file). However it appears more complicated as there are also files that are non-seekable where tell returns a valid result:
https://github.com/asdf-format/asdf/issues/1552

I will work on addressing hopefully all 3 of these in this PR. My current thoughts are:
- modify generic_io to catch the `Illegal seek` returned from tell and instead return None
- modify the block index generation to use tell for non-seekable files and to not write out an index if it contains all 0s (because of the windows bug which has wider implications) or all Nones

braingram replied:
I believe I addressed the non-seekable file write issues here:
684ccf375ab6dfb99b61a625d2a53b7f30a62082
the above commit also adds tests for the various odd `tell` results.

braingram commented on asdf/_block/manager.py:
```diff
@@ -0,0 +1,589 @@
+        self._items.append(blk)
+        self._by_data.assign_object(blk._data, index)
+        self.assign_object_by_index(obj, index)
+        return index
```
Yeah this is a bit messy. Originally `LinearStore` was used for more than just `WriteBlocks`. I'm certain this organization can be made clearer and I appreciate your feedback on this (as my head has been deep in this code for a while which makes it hard to judge the readability and completeness of the documentation).

I removed `LinearStore` and moved the functionality into `WriteBlocks`:
ae911d8b36261e7444c6fae5b49ddc4672240690
which I also cleaned up and added some docs to hopefully describe the layout:
b65023a19b6fdf6c2af8351800ef42b1ee33c523

Does this help to explain the new layout and what the Store is used for?

braingram commented on asdf/_block/store.py:
```diff
@@ -0,0 +1,128 @@
+            del self._by_id[object_id]
+
+
+class LinearStore(Store, collections.abc.Sequence):
```
I ended up removing `LinearStore`. See https://github.com/asdf-format/asdf/pull/1537#discussion_r1212142870 for more details.

braingram commented:
WelDX errors are due to a new version of pint (0.22). In brief, `pint.Quantity` now appears as `pint.registry.Quantity` when the class name is inspected. This, combined with asdf matching by string when a converter type is provided as a string resulted in the failures. See this PR for a fix:
https://github.com/BAMWelDX/weldx/pull/880

braingram commented on asdf/yamlutil.py:
```diff
@@ -217,12 +218,16 @@ def custom_tree_to_tagged_tree(tree, ctx, _serialization_context=None):
     if _serialization_context is None:
         _serialization_context = ctx._create_serialization_context()
```
I think so.

Looking at how `SerializationContext` is passed around I think that the type of operation could be defined at time of creation and that all uses in arguments are private.
Outside of the tests, the only creation of a `SerializationContext` instance occurs during `AsdfFile._create_serialization_context`.
https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/asdf.py#L1668
These instances are passed into/used in many different functions (ignoring the legacy extension code) including:
`AsdfFile._write_tree`: https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/asdf.py#L990-L994
Several uses in `asdf.yamlutil`
https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/yamlutil.py#L211
https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/yamlutil.py#L276
https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/yamlutil.py#L361
One use in `asdf.schema.get_validator`
https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/schema.py#L515

I added a commit that adds an argument to `AsdfFile._create_serialization_context` and changed the subclasses in `asdf._serialization_context` (deleting `_Operation`, renaming `_Serialization` to `WriteBlocksContext` and `_Deserialization` to `ReadBlocksContext` and added `asdf._serialization_context.create` to simplify creation of contexts). The details can be found in the commit:
3eb2cf23f68aaf8bd30b51688d52e1a3a0fbabbf

braingram commented on asdf/yamlutil.py:
```diff
@@ -217,12 +218,16 @@ def custom_tree_to_tagged_tree(tree, ctx, _serialization_context=None):
+    sctx = _Serialization(_serialization_context)
+
+    extension_manager = sctx.extension_manager
+    version_string = ctx.version_string
```
Thanks! Fixed in: 4db9c91be6cf7e06c47b2e1ee0846092cb9bc831

braingram commented on asdf/yamlutil.py:
```diff
@@ -241,25 +246,46 @@ def _convert_obj(obj):
 
+    cfg = config.get_config()
+    convert_ndarray_subclasses = cfg.convert_unknown_ndarray_subclasses
+    converters = {}
```
The `converter` caching is a minor performance improvement (more details in this comment: https://github.com/asdf-format/asdf/pull/1537#issuecomment-1551755705).
The `convert_ndarray_subclasses` copy is probably inconsequential but does avoid recalling the `AsdfConfig` property getter for every object.

braingram commented on asdf/_block/manager.py:
```diff
@@ -0,0 +1,675 @@
+    """
+
+    def __init__(self, blocks=None):
+        super().__init__()
```
The tests pass without this call to `super().__init__()` so I removed it in:
1c28b48f40c0bc7f14db39e77fccf0c6c57534b1

braingram commented:
dkist failures look unrelated to this PR and instead likely due to the more recent runs testing with 'globus==3.20.x' instead of '3.19.x' and this draft dkist PR:
https://github.com/DKISTDC/dkist/pull/198

WilliamJamieson commented:
> Tests that correspond to a specific issue are organized in `_tests/_issues`. Each issue gets it's own file `test_1520.py` with a function `test_1520` with a docstring linking to the github issue and code that closely reflects any minimal test case in the issue.

I vehemently disagree with this approach. Test names should be descriptive of what they are testing for. There is no way for anyone to have a clue what a test is about from the failure report from `pytest` as an issue number is basically arbitrary.

Yes the docstring for the test should link back to the issue on GitHub, and sure the docstring should contain some sort of description of what is failing. However, I should not have to actually look at the source code to have any idea of what could be breaking. This will make contributing to the source code more and more hostile to new contributors as the number of issue related tests rises.

WilliamJamieson commented on asdf/_tests/_issues/test_1013.py:
```diff
@@ -0,0 +1,45 @@
+import asdf
+
+
+def test_1013(tmp_path):
```
As noted in https://github.com/asdf-format/asdf/pull/1537#issuecomment-1618994330. I strongly disagree with naming tests after issues. I also disagree with your organization method of each issue being its own module. The reason being that the test modules are "supposed" to be some reasonable categorization of like tests. For example, the `test_manager` module should contain most of the tests which will "smoke" out any issues related to the "manager". This way when working on a specific section of the code base the developer can iterate much faster against the tests categorized to be related to what they are doing. This becomes ever more important the longer the test suite takes to run.

This naming scheme both destroys this categorization AND as I already commented, makes what tests are supposed to be testing extremely opaque. At the very least it should not be both, and I think the test name is more important than the module name. However, if you want to go this way I believe you should introduce categorical test markers to so that the tests can be reduced to specific categories.

Since the "issue" categorization is useful I think things should be organized via test classes instead. By this I mean:
```python
class Test1013:
    """documentation linking to #1013 and a basic description"""

    # Specifically named tests concerning the solution to #1013
    ...
```
These test classes can then be incorporated into the categorical modules that already exist or in new ones as needed. By doing this you largely cover all the major concerns:

1. Issues being directly linked to tests.
2. Descriptive test names.
3. Test categorization.

I do grant that the current test categorization/modules do leave much to be desired, and so I fully support reorganizing those and their tests into better "categories".

WilliamJamieson commented:
> Fixes or work towards fixes for many issues are included in this PR
> 
> * [controlling array storage in to_yaml_tree methods #1013](https://github.com/asdf-format/asdf/issues/1013) a test was added for a possible fix for this, if we decide that `SerializationContext` should have the same public `set_array_storage` methods as `AsdfFile` than this issue could be solved
> * fixes [A failed update can corrupt the file #1520](https://github.com/asdf-format/asdf/issues/1520)
> * fixes [update corrupts stream data #1523](https://github.com/asdf-format/asdf/issues/1523)
> * fixes [array storage settings are inconsistently propagated to new AsdfFile and can produce a file that cannot be opened #1524](https://github.com/asdf-format/asdf/issues/1524) by not copying block options between AsdfFiles
> * fixes [Rewriting a file with external blocks fails if arrays are not first accessed #1526](https://github.com/asdf-format/asdf/issues/1526)
> * [Calling update with memmapped data can create invalid data in memmap views #1530](https://github.com/asdf-format/asdf/issues/1530) a test was added and xfailed as this may require subclassing ndarray
> * fixes [Support file truncating on windows #1534](https://github.com/asdf-format/asdf/issues/1534)
> * fixes [ASDF unable to read empty inline array. #1538](https://github.com/asdf-format/asdf/issues/1538)
> * [Seek and read from closed MemoryIO #1539](https://github.com/asdf-format/asdf/issues/1539) a test was added and xfailed, this will require possibly major changes to `generic_io`
> * fixes [ASDF writes but fails to read inline structured array #1540](https://github.com/asdf-format/asdf/issues/1540)
> * fixes [Block checksums are only checked for first block if a block index is present #1541](https://github.com/asdf-format/asdf/issues/1541)
> * fixes [ASDF fails to write blocks to non-seekable file  #1542](https://github.com/asdf-format/asdf/issues/1542)
> '

If any of these fixes is largely independent of the moving of ndarray conversion to a converter, then they should each be in their own distinct PR.

pllim reviewed (commented):
Drive by comment: I don't think naming tests after issue number is a good idea.

No one memorizes issue numbers, so at a glance I have no idea what you are supposed to be testing without have to dig into your GitHub issues.

What if an issue need to be tested across different modules? What if a future issue is opened against this past issue about the same thing? A test should be named based on its intention, not just mere numbers.

And what if you move the code from GitHub to GitLab in the future? Or merge this into a different repo or split this repo up? Then the numbers don't mean anything anymore. When STScI moved from SVN to Git, all our names/comments referring to Trac tickets became meaningless.

pllim commented:
Here is an example from the past: https://github.com/spacetelescope/pysynphot/blob/387eb5a4f90a031b28400f27ac34469606d50e5b/pysynphot/test/test_box.py#L15

```python
def test_ticket114():
    bp = Box(500, 10, waveunits='nm')
    assert str(bp.waveunits) == 'nm'  # test units
    assert bp.sample(500) == 1  # test inside
```

I have no idea what Ticket 114 is. If I ever have to update this test, I can never be sure if I have changed its original intention or not. It is definitely not the 114 on GitHub:

* https://github.com/spacetelescope/pysynphot/pull/114

braingram commented on asdf/_tests/_regtests/test_1525.py:
```diff
@@ -0,0 +1,29 @@
+    af.set_array_storage(arr, "external")
+    af.write_to(fn)
+
+    for copy_arrays in (True, False):
```
Thanks! I blindly copied this from the issue and changing it to a parametrization is cleaner. Added the parametrization here: 5dad8d0795d57008193d37129e9dd46f2399a721

braingram commented on asdf/_tests/_regtests/test_1530.py:
```diff
@@ -0,0 +1,35 @@
+        np.testing.assert_array_equal(a, af["a"])
+        np.testing.assert_array_equal(b, af["b"])
+        assert False
+        # np.testing.assert_array_equal(va, ov)  # segfault
```
Fingers crossed this PR will allow us to more easily turn `NDArrayType` into a `ndarray` subclass to allow us to test things like `__array_priority__` to force views to be `NDArrayType` to keep the close check and other options for preventing segfaults.

braingram commented on asdf/commands/edit.py:
```diff
@@ -130,23 +131,42 @@ def write_edited_yaml_larger(path, new_content, version):
+                    this_size = min(block_size, n_bytes - offset)
+                    fd.write(original_fd.read(this_size))
+
+                # update index
```
I did find that a bit odd that `asdftool edit` would add an index to a file that didn't previously had one (unless it has a streaming block) only if the tree expanded into the blocks. I don't quite recall if at the time I considered this a 'bug' in edit and changed it to only add an index if the original file had one. I would argue that this makes more sense then sometimes adding an index to a file that didn't previously have one.

It shouldn't be a problem to change it back to having it always generate an index. However this would require parsing the last block header to make sure an index isn't added for a file with a streaming block.

Shall we consider the adding of a block index a bug and keep the current behavior in this PR or should I update it to parse the last block header and add an index to a file if the tree expands into the blocks?

braingram commented on asdf/yamlutil.py:
```diff
@@ -264,19 +267,40 @@ def _convert_obj(obj):
+            )
+            converter = extension_manager._get_converter_for_subtype(typ)
+            converters_cache[typ] = lambda obj, _converter=converter: _convert_obj(obj, _converter)
+            return _convert_obj(obj, converter)
```
Great idea! I updated the code to use your simplified code: 430cb30727d24f865597bd4d639575e4353d7a4f

braingram commented on asdf/_tests/_regtests/test_1520.py:
```diff
@@ -0,0 +1,34 @@
+import asdf
+
+
+def test_failed_update_corrupts_file(tmp_path):
```
Yes and no :)

The original issue was discovered while testing update and compressed data. If during an update a compressed block outgrew it's previous size the update would fail and the original file would be left in a 'corrupt' state where the tree did not match the blocks.

Ending up corrupting the original file always seems like a risk when calling 'update' (what would happen if halfway through writing the tree your computer lost power?).

If I recall correctly, I added the regtest prior to a working 'update' with the new block management code. While working on update the conditions that triggered the failed update were fixed (with the code in this PR compressed blocks can grow during an update and no error occurs). I ended up leaving in the test and modifying it to reflect that the update no longer fails.

I removed `test_1520` (but I think we should update and close the issue with this PR since the code that previously failed is fixed by this PR). I also added a new test `test_update_compressed_blocks` with some notes about the above history and to test that updating compressed data (causing it to increase size on disk) no longer causes update to fail: 590c9a35fa2bc7688801c297bd00ab59eb17117f

braingram commented on pyproject.toml:
```diff
@@ -38,7 +38,7 @@ all = [
 ]
 docs = [
-  "sphinx-asdf>=0.1.4",
+  "sphinx-asdf @ git+https://github.com/braingram/sphinx-asdf.git@immutable_block_manager",
```
Thanks! If the overall internal API changes in this PR look good then we could work on getting the corresponding sphinx-asdf PR merged first: https://github.com/asdf-format/sphinx-asdf/pull/64
sphinx-asdf would then need to be released with the changes before switching this line to `>=0.3` (or whatever version includes the above linked sphinx-asdf PR.

braingram commented on asdf/extension/_serialization_context.py:
```diff
@@ -1,5 +1,8 @@
+from asdf.extension._extension import ExtensionProxy
 
 
 class SerializationContext:
```
I think that's a great idea. I want to say that's enough of a change that it could go in a separate PR so I'll open an issue to track it: https://github.com/asdf-format/asdf/issues/1626

braingram commented on asdf/yamlutil.py:
```diff
@@ -310,7 +334,10 @@ def _walker(node):
 
         if extension_manager.handles_tag(tag):
             converter = extension_manager.get_converter_for_tag(tag)
+            _serialization_context.assign_object(None)
```
Good catch. It shouldn't. I removed the line in: 290b43d5393588b90770f6e51bc99d136730e6f7

braingram commented on asdf/_block/reader.py:
```diff
@@ -0,0 +1,259 @@
+                msg = "Invalid block magic"
+                raise OSError(msg)
+            blocks[index].load()
+    except (OSError, ValueError):
```
Great suggestion! I added warnings here: e4b2b5edcf20f4e5562966a643babfdb4741721f

A few tests required updates (as the tests expected block index failures). Interestingly the `block0.asdf` and `block1.asdf` data files used in `test_diff_block` both had invalid block indices. I deleted the block index in each file as it doesn't seem important for the test.

braingram replied:
I made one more small change to this after seeing the sunpy and dkist errors (due to using test data with invalid block indices). I changed the warnings to a new `AsdfBlockIndexWarning`: 949d1d1bb4b648ca9df1fc4175dad0ef0755ea15
This should allow them to ignore the new warning without ignoring all `AsdfWarning`s.
On a related note (not to address in this PR), perhaps we should rename all of these to remove the extraneous `Asdf` prefix?

braingram commented on asdf/core/_converters/ndarray.py:
```diff
@@ -0,0 +1,184 @@
+            options = ctx._blocks.options.get_options(data)
+
+        # possibly override options based on config settings
+        with config_context() as cfg:
```
Good catch! I updated the code here: c86b5b53057d62c155a09c29b92a0297e18e8cfa

braingram commented on asdf/generic_io.py:
```diff
@@ -380,7 +380,6 @@ def seek(self, offset, whence=0):
             file`s end).
         """
         result = self._fd.seek(offset, whence)
-        self.tell()
```
Not that I'm aware of. Is there a side effect of tell?

Looks like it's been hiding here for 9 years:
https://github.com/asdf-format/asdf/blame/main/asdf/generic_io.py#L382-L384

braingram commented on asdf/extension/_serialization_context.py:
```diff
@@ -77,85 +81,179 @@ def _extensions_used(self):
-        block_index : int
-            The index of the block to associate with the key
+        key : BlockKey, optional
+            BlockKey generated using self.generate_block_key. Only
```
One possibility might be if an object wants to pass a block reference to another object. I also can't think of a realistic reason to do this. I don't think anything should prevent the use of a key for a single block (and some zarr arrays might end up with 1 block if they're not fully populated).

braingram commented:
> A final smattering of not-very-consequential comments. Thanks for bearing with me!

"final" :-O :)

Thanks for the suggestions and questions! They improved the code.

I do see a few new errors in sunpy and dkist because of the warnings from invalid block indicies that will need some investigation.

braingram commented:
I updated and release sphinx-asdf and updated pyproject.toml (setting a lower pin and removing the dev requirement) so now the docs build with the pypi version of sphinx asdf.

I've opened PR for:
- stdatamodels to convert all fitsrec to arrays prior to passing them to asdf for validation or serialization: https://github.com/spacetelescope/stdatamodels/pull/205 (I took a brief stab at completely removing fitsrec but there is at least one use of it for storing units and numerous uses of things like fields and the case insensitivity of names which will need to be sorted): https://github.com/spacetelescope/stdatamodels/pull/205
- asdf-astropy to add a converter/tag/schema for `NdarrayMixin`: https://github.com/astropy/asdf-astropy/pull/200
- dkist to update files with incorrect block indexes: https://github.com/DKISTDC/dkist/pull/296
